### DISCUSSION TITLE
Serializing Structs Containing Custom Type Vectors to JSON with Macros

### DISCUSSION THREAD

**Discussion by tomlin7** [nlohmann/json#2844] [Q&A] [answered]
Let me explain my case,
I have a struct named Shirt

```cpp
struct Shirt
{
    std::string color;
    int size;
};
```
And I have another struct named Person, which uses a vector of Shirt
```cpp
struct Person
{
    std::string name;
    int age;

    // The issue comes here.
    std::vector<Shirt> shirts;
};
```
here, I need help on converting this Person struct to JSON.
with the help of docs, I used the `NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE` macro for it

```cpp
struct Person
{
    std::string name;
    int age;
    std::vector<Shirt> shirts;

    NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(Person, name, age, shirts)
};
```
And I'm getting an error at this point.
![image](https://user-images.githubusercontent.com/70792552/123763539-86494500-d889-11eb-8a29-a2a0857012be.png)

```
Severity	Code	Description	Project	File	Line	Suppression State
Error	C2679	binary '=': no operator found which takes a right-hand operand of type 'const std::vector<main::Shirt,std::allocator<main::Shirt>>' (or there is no acceptable conversion)	playground	D:\projects\cpp\playground\playground\playground.cpp	226	
```
A sample i tried is following
```cpp
std::vector<Shirt> shirts = { { "green",10 } };
Person person{ "ned", 25, shirts };

json j;
person.to_json(j, person);
```

The result I'm expecting (the value of `j` in this case) from the serialization is the following
```json
{
    "age": 25,
    "name": "ned",
    "shirts": [
        {
            "color": "green", 
            "size": 10
        }
    ]
}
```

---

What am I doing wrong at this point?
If so, how can I achieve the result I'm expecting?
🙏🏻

**Comment by gregmarr** [comment]
> There are two macros to make your life easier as long as you (1) want to use a JSON object as serialization and (2) want to use the member variable names as object keys in that object:
> 
> NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(name, member1, member2, ...) is to be defined inside of the namespace of the class/struct to create code for.
> NLOHMANN_DEFINE_TYPE_INTRUSIVE(name, member1, member2, ...) is to be defined inside of the class/struct to create code for. This macro can also access private members.

You are using `NON_INTRUSIVE`, which goes in the namespace of the class, but putting it inside the class.  You either want to use `INTRUSIVE` or move it outside the class.  You also need to add the macro for the `Shirt` class.

**Reply by tomlin7** [reply]
Keeping the `Person` as `NON_INTRUSIVE` and Adding an `INTRUSIVE` macro to the `Shirt` helped me achieve what I want 👍🏻 Thank you for helping me out :)
Also, my structs were defined in the `main()` which was causing the error, so I moved them out.

**Comment by tomlin7** [comment, marked as answer]
- I was defining the structs in the main function, which was causing the main issue.
- with the help of what @gregmarr said, I added the macro to the `Shirt` as well, so now it works as I expected.

---

Here's the complete code
```cpp
#include <iostream>
#include <nlohmann/json.hpp>

using json = nlohmann::json;

struct Shirt
{
    std::string color;
    int size;

    NLOHMANN_DEFINE_TYPE_INTRUSIVE(Shirt, color, size)
};

struct Person {
    std::string name;
    int age;
    std::vector<Shirt> shirts;

    NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE(Person, name, age, shirts)
};

int main()
{
    std::vector<Shirt> shirts = { { "green",10 } };
    Person person{ "ned", 25, shirts };

    json j;
    person.to_json(j, person);

    std::cout << j;

    return 0;
}
```

And the output
```
{"age":25,"name":"ned","shirts":[{"color":"green","size":10}]}
```

---

## I went through some experiments with these macros, results are following

#### Keeping `NON_INTRUSIVE` for `Person` and `INTRUSIVE` for `Shirt` (this one worked)
> ![image](https://user-images.githubusercontent.com/70792552/123843063-15307e80-d8d7-11eb-86c6-38153af8fb61.png)

output
> ![image](https://user-images.githubusercontent.com/70792552/123843884-faaad500-d8d7-11eb-9bdd-f98252f9318b.png)

---

#### Having `INTRUSIVE` for both `Person` and `Shirt`
> ![image](https://user-images.githubusercontent.com/70792552/123843020-05189f00-d8d7-11eb-8d0c-40256a24ff6d.png)

Error
> ![image](https://user-images.githubusercontent.com/70792552/123843964-10b89580-d8d8-11eb-8db0-bbe32c356caa.png)

---

#### Having `NON_INTRUSIVE` for both `Person` and `Shirt`
> ![image](https://user-images.githubusercontent.com/70792552/123843442-853f0480-d8d7-11eb-8fd8-ec783d517b1d.png)

Error
> ![image](https://user-images.githubusercontent.com/70792552/123843712-d222db00-d8d7-11eb-82f4-682e4cdce66d.png)


---

Thanks to @gregmarr for helping me out!

**Reply by gregmarr** [reply]
https://godbolt.org/z/cWEEohKhz

You want this:

    j = person;

not this:

    person.to_json(j, person);

**Reply by tomlin7** [reply]
That's more convenient! The last time I tried I was getting a type conversion error, but this is working as per the compiler explorer project, will check it out, Thanks again 👍🏻
